src/service/position-observer/position-observer-impl.js

```diff
@@ -0,0 +1,243 @@
+      this.onScrollHandler_();
+    }));
+    this.unlisteners_.push(this.viewport_.onResize(() => {
+      this.onResizeHandler_();
+    }));
+  }
+
+  /**
+   * Callback function that gets called when unobserve last observed element.
+   * @private
+   */
+  stopCallback_() {
+    while (this.unlisteners_.length) {
+      const unlisten = this.unlisteners_.pop();
+      unlisten();
+    }
+  }
+
+  /**
+   * @param {!Element} element
+   * @param {PositionObserverFidelity} fidelity
+   */
+  changeFidelity(element, fidelity) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+      if (entry.element == element) {
+        entry.fidelity = fidelity;
+        if (fidelity == PositionObserverFidelity.HIGH) {
+          entry.turn = 0;
+        }
+        return;
+      }
+    }
+    dev().error(TAG, 'cannot change fidelity on unobserved element');
+  }
+
+  /**
+   * This should always be called in vsync.
+   * @param {boolean=} opt_force
+   * @visibleForTesting
+  */
+  updateAllEntries(opt_force) {
+    for (let i = 0; i < this.entries_.length; i++) {
+      const entry = this.entries_[i];
+
+      if (opt_force) {
+        this.updateSingleEntry_(entry);
+        continue;
+      }
+
+      if (entry.turn == 0) {
+        this.updateSingleEntry_(entry);
+        entry.turn = (entry.fidelity == PositionObserverFidelity.LOW) ?
+            LOW_FIDELITY_FRAME_COUNT : 0;
+      } else {
+        entry.turn--;
+      }
+    }
+  }
+
+  /**
+   * To update the position of single element when it is ready.
+   * Called when updateAllEntries, or when first observe an element.
+   * @param {!./position-observer-entry.PositionObserverEntry} entry
+   * @private
+   */
+  updateSingleEntry_(entry) {
+    const viewportSize = this.viewport_.getSize();
+    const viewportBox =
+        layoutRectLtwh(0, 0, viewportSize.width, viewportSize.height);
+    this.viewport_.getBoundingRectAsync(entry.element).then(elementBox => {
+      entry.trigger(
+      /** @type {./position-observer-entry.PositionInViewportEntryDef}*/ ({
+        positionRect: elementBox,
+        viewportRect: viewportBox,
+        relativePos: '',
+      }));
+    });
+  }
+
+  /**
+   * Handle viewport scroll event
+   * @private
+   */
+  onScrollHandler_() {
+    this.boundStopScroll_();
+    this.inScroll_ = true;
+    if (!this.measure_) {
+      this.schedulePass_();
+    }
+  }
+
+  /**
+   * Handle viewport resize event
+   * @private
+   */
+  onResizeHandler_() {
+    this.updateAllEntries(true);
+  }
+
+  /**
+   * Update all entries during scroll
+   * @private
+   */
+  schedulePass_() {
+    // TODO (@zhouyx, #9208):
+    // P1: account for effective fidelity using this.effectiveFidelity
+    // P2: do passes on onDomMutation (if available using MutationObserver
+    // mostly for in-a-box host, since most DOM mutations are constraint to the
+    // AMP elements).
+    this.updateAllEntries();
```

Comment:
Shouldn't this wait for a measure?

Reply:
`#updateAllEntries()` call `#updateSingleEntries()` then call `#Viewport.getBoundingRectAsync()`. which returns rect in next vsync measure.